### PR TITLE
Added Capability Construct  and strict schema for component style

### PR DESCRIPTION
**Pull request by codeSafari10**
**Notes for Reviewers**

### Description:

This PR introduces a significant refactoring of our component capabilities schema to improve scalability, flexibility, and clarity. The changes aim to flatten the capability structure and introduce a more generic, extensible approach to defining component capabilities to achieve that we are creating a new construct called `capability`

Key changes:

1. Replaced the hierarchical object structure with a new, flattened capability construct.
2. Introduced a new schema for capabilities with the following key fields:
   - `kind`: High-level category of the capability
   - `type`: More specific categorization within each kind
   - `subType`: (Optional) Most granular level of capability definition
   - `state`: Specifies whether the capability applies to "declaration" or "instance" state of an entity
   - `status`: Indicates if the capability is "enabled" or "disabled"
   - Additional fields: `displayName`, `description`, `key` for more detailed capability information

3. Defined enum variants for `kind`  , while type and subtype are flexible  to cover a wide range of capabilities without frequent schema changes.

### Benefits:

- Increased scalability: New capabilities can be added without changing the core schema.
- Improved clarity: Capabilities are now more precisely categorized and described.
- Enhanced flexibility: The new structure allows for more detailed capability definitions.
- Better organization: Capabilities are now consistently structured across different component states.

### Example:

Old format:
```json
"designer": {
  "capabilities": {
    "config": { "type": "boolean" },
    "label": { "type": "boolean" }
  }
}
```

New format:
```json
{
  "kind": "configuration",
  "type": "edit",
  "subType": "component-config",
  "state": ["declaration"],
  "status": "enabled",
  "displayName": "Edit Configuration",
  "description": "Allows configuration of the component",
}
```

## Example Component Capabilities 
```
{
  kind : "mutate"
  type :  "configuration"
  subType : "config" | "labels" | "annotations" 
  description : "Gives the ability to edit a components configuration" 
}   // drives the configuration tooltip in extensions

{
  kind: "mutate"
  type: "style" 
  subtype : "shape" | "label"  ...  ( absence of a property means it is allowed ) 
} 

{
  kind : "interaction" 
  type: "graph"
  subtype: "allowsCompoundDragAndDrop" | "bubbleEvents" | "lock"  | "mouse-events"
} 

{
  kind: "view" 
  type: "configuration"
  subtype :  "config" | "labels" | "annotations"   // whether a readonly view is supported 
} 

{
  kind: "view" 
  type: "graph"
  subtype :  "labels" | "children" 
} 

```

### Next Steps:

- Update all existing component definitions to use the new capability schema.
- Refactor related code to work with the new capability structure.

---

Would you like me to modify or expand on any part of this pull request description?

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

**Comment by aabidsofi19** [comment]
i tried to make the kind generic so the capabilty schema becomes generic for all entities .  i think there is still some room for making them more generic ,less overlapping . and maybe reducing the number . i guess some of the kinds right now then might go into type .

**Comment by aabidsofi19** [comment]
@codeSafari10  what do you think ?

**Comment by codeSafari10** [comment]
reduced kind to : 

```
    "kind": {
      "$ref": "../core.json#/definitions/inputString",
      "description": "Kind of the capability",
      "anyOf": [
        {
          "const": "configuration",
          "description": "For capabilities related to entities configuration  like config, labels , annotations,etc"
        },
        {
          "const": "visualization",
          "description": "For capabilities related to visualization of entities like labels , styles"
        },
        {
          "const": "lifecycle",
          "description": "For capabilities related to lifecycle mamangement of entities like deploy, undeploy ,start, stop, restart, etc"
        },

        {
          "const": "action",
          "description": "Catch all for capabilities related to actions on entities like delete,run, execute, etc"
        }
      ]
    },

```

**Comment by codeSafari10** [comment]
@aabidsofi19  implementation specific list of component capabilities : ```
const ComponentCapabilities: CapabilityIdentifier[] = [
  {
    kind: "configuration",
    type: "edit",
    subType: "config"
  },
  {
    kind: "configuration",
    type: "management",
    subType: "gitops"
  },
  {
    kind: "visualization",
    type: "edit",
    subType: "style"
  },
  {
    kind: "visualization",
    type: "edit",
    subType: "shape"
  },
  {
    kind: "visualization",
    type: "edit",
    subType: "resize"
  },
  {
    kind: "visualization",
    type: "interaction",
    subType: "compoundDragAndDrop"
  }
];
```

**Comment by codeSafari10** [comment]
ref to the key schema is missing here . i dont know if the openapi key schema works for a general key construct

**Comment by leecalcote** [comment]
@Jougan-0, you're going to want to get your mind wrapped around this. // @Yashsharma1911 @humblenginr

**Comment by leecalcote** [comment]
> @Jougan-0, you're going to want to get your mind wrapped around this. // @Yashsharma1911 @humblenginr

And quickly pair a PR updating the Model Generator to reflect these changes.

File: schemas/constructs/v1alpha1/capability.json [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      "description": "Name of the capability in human-readible format."
+    },
+    "description": {
+      "type": "string",
+      "description": "A written representation of the purpose and characteristics of the capability."
+    },
+
+    "kind": {
+      "$ref": "../core.json#/definitions/inputString",
+      "description": "Kind of the capability",
```
Comment (leecalcote) [inline schemas/constructs/v1alpha1/capability.json:32]:
```suggestion
      "description": "Top-level categorization of the capability",
```

File: schemas/constructs/v1alpha1/capability.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      "items": {
+        "type": "string",
+        "enum": ["declaration", "instance"]
+      }
+    },
+
+    "status": {
+      "type": "string",
+      "description": "Status of the capability",
+      "enum": ["enabled", "disabled"]
```
Comment (leecalcote) [inline schemas/constructs/v1alpha1/capability.json:78]: Good use of `status`. This property should become more or less a universal property in all Meshery constructs.

File: schemas/constructs/v1alpha1/capability.json [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    },
+
+    "metadata": {
+      "type": "object",
+      "description": "Metadata contains additional information associated with the capability.",
```
Comment (leecalcote) [inline schemas/constructs/v1alpha1/capability.json:83]:
```suggestion
      "description": "Metadata contains additional information associated with the capability. Extension point.",
```

File: schemas/constructs/v1alpha3/relationship.json [hunk 4]
```diff
@@ -53,6 +42,13 @@
         "description": {
           "$ref": "../core.json#/definitions/inputString",
           "description": "Characteristization of the meaning of the relationship and its relevance to both Meshery and entities under management."
+        },
+        "capabilities": {
```
Comment (leecalcote) [inline schemas/constructs/v1alpha3/relationship.json:46]: Given our discussion regarding probability of change here, I'll suggest that we promote capabilities as a top-level property.

File: schemas/constructs/v1alpha3/relationship.json [hunk 5]
```diff
@@ -53,6 +42,13 @@
         "description": {
           "$ref": "../core.json#/definitions/inputString",
           "description": "Characteristization of the meaning of the relationship and its relevance to both Meshery and entities under management."
```
Comment (leecalcote) [inline schemas/constructs/v1alpha3/relationship.json:44]:
```suggestion
          "description": "Characterization of the meaning of the relationship and its relevance to both Meshery and entities under management."
```

File: schemas/constructs/v1alpha3/relationship.json (hunk 4 above)
Comment (leecalcote) [reply schemas/constructs/v1alpha3/relationship.json:46]: Leaving these optional. ✅

File: schemas/constructs/v1beta1/component.json [hunk 6]
```diff
@@ -43,16 +40,225 @@
...
+
+              "width": {
+                "type": "number",
+                "description": "The width of the node\u2019s body or the width of an edge\u2019s line."
+              },
+              "height": {
+                "type": "number",
+                "description": "The height of the node\u2019s body"
+              },
+              "shape": {
```
Comment (leecalcote) [inline schemas/constructs/v1beta1/component.json:89]: Perhaps, `shape` could be a core definition.

File: schemas/constructs/v1alpha1/capability.json [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      "description": "A written representation of the purpose and characteristics of the capability."
+    },
+
+    "kind": {
+      "$ref": "../core.json#/definitions/inputString",
+      "description": "Kind of the capability",
+      "anyOf": [
+        {
+          "const": "action",
+          "description": "For capabilities related to executing actions on entities"
```
Comment (leecalcote) [inline schemas/constructs/v1alpha1/capability.json:36]:
```suggestion
          "description": "For capabilities related to executing actions on entities. Example: initiate log streaming on a Pod. Example: initiate deployment of a component."
```

File: schemas/constructs/v1alpha1/capability.json [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        },
+        {
+          "const": "mutate",
+          "description": "For capabilities related to mutating an entity"
+        },
+
+        {
+          "const": "interaction",
+          "description": "Catch all for capabilities related to interaction with entities"
```
Comment (leecalcote) [inline schemas/constructs/v1alpha1/capability.json:45]:
```suggestion
          "description": "Catch all for capabilities related to interaction with entities. Example: the ability for a component to be dragged and dropped. Example: supports event bubbling to parent components. "
```

**Comment by leecalcote** [comment]
A foreseeable possibility in the future is that Capabilities might have Selectors in the same way that Relationships have Selectors. Thank you for this discussion on this concept today, @codeSafari10.

File: schemas/constructs/v1alpha1/capability.json [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        }
+      ]
```
[the remark this answers is not in the document]
Comment (leecalcote) [reply schemas/constructs/v1alpha1/capability.json:47]: @MUzairS15, what do you think? Overkill?

File: schemas/constructs/v1alpha1/capability.json [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      "$ref": "../core.json#/definitions/inputString",
+      "description": "Most granular unit of capability classification. The combination of Kind, Type and SubType together uniquely identify a Capability."
+    },
+
+    "key": {
+      "$ref": "../core.json#/definitions/inputString",
+      "description": "Key that backs the capability."
+    },
+
+    "state": {
```
[the remark this answers is not in the document]
Comment (MUzairS15) [reply schemas/constructs/v1alpha1/capability.json:65]: +1

File: schemas/constructs/v1alpha1/capability.json (hunk 2 above)
Comment (MUzairS15) [reply schemas/constructs/v1alpha1/capability.json:78]: Add a default value as "enabled"

File: schemas/constructs/v1alpha1/capability.json [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+  "properties": {
+    "schemaVersion": {
+      "$ref": "../core.json#/definitions/versionString",
+      "description": "Specifies the version of the schema to which the capability definition conforms."
+    },
+    "version": {
+      "$ref": "../core.json#/definitions/semverString",
+      "description": "Version of the capability definition.",
+      "minLength": 1,
```
Comment (MUzairS15) [inline schemas/constructs/v1alpha1/capability.json:18]: Shouldn't this be standard?
May be add in the semverString schema?

File: schemas/constructs/v1alpha1/capability.json [hunk 12]
```diff
@@ -0,0 +1,88 @@
+{
+  "$id": "https://schemas.meshery.io/capability.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "description": "Meshery manages entities in accordance with their specific capabilities. This field explicitly identifies those capabilities largely by what actions a given component supports; e.g. metric-scrape, sub-interface, and so on. This field is extensible. Entities may define a broad array of capabilities, which are in-turn dynamically interpretted by Meshery for full lifecycle management.",
+  "additionalProperties": false,
+  "type": "object",
+
+  "required": ["schemaVersion", "version", "displayName", "kind", "type", "state", "status"],
```
Comment (MUzairS15) [inline schemas/constructs/v1alpha1/capability.json:8]: kind, type and subtype together should uniquely identify a capability. Hence subType should be required

File: schemas/constructs/v1alpha1/capability.json [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      "description": "Classification of capabilities. Used to group capabilities similar in nature."
+    },
+
+    "subType": {
+      "$ref": "../core.json#/definitions/inputString",
+      "description": "Most granular unit of capability classification. The combination of Kind, Type and SubType together uniquely identify a Capability."
+    },
+
+    "key": {
+      "$ref": "../core.json#/definitions/inputString",
```
Comment (MUzairS15) [inline schemas/constructs/v1alpha1/capability.json:61]: As combination of kind, type and subType together uniquely identify a capability, so do we need this?

We can have key identified / discerned as `kind_type_subType`.

File: schemas/constructs/v1alpha1/capability.json (hunk 12 above)
Comment (leecalcote) [reply schemas/constructs/v1alpha1/capability.json:8]: @MUzairS15, my guess is the thinking here is that it's not always necessary to have this level specificity in order to identify a capability. Given this, that begs the idea that `subtype` is optional.

If we can put some forethought into a proposed set of classifications (a set of examples of how kind, type, subtype can be made to always be present), that exercise will lend itself to making it comfortable to require subtype.

File: schemas/constructs/v1alpha1/capability.json (hunk 13 above)
Comment (leecalcote) [reply schemas/constructs/v1alpha1/capability.json:61]: @MUzairS15 this is more or less a placeholder for future incorporation of permissions.

File: schemas/constructs/v1beta1/component.json [hunk 14]
```diff
@@ -29,42 +29,113 @@
...
+              "description": "The y-coordinate of the node."
+            }
+          }
         },
-        "genealogy": {
+
+        "body-text": {
           "type": "string",
-          "description": "Genealogy represents the various representational states of the component."
+          "description": "The text to display for an element\u2019s body. Can give a path, e.g. data(id) will label with the elements id"
```
Comment (leecalcote) [inline schemas/constructs/v1beta1/component.json:84]: I wonder if use of unicode places any undue burden on clients that might display this string literally. Hopefully, most clients are interpreting and displaying the apostrophe.

File: schemas/constructs/v1beta1/component.json [hunk 15]
```diff
@@ -89,28 +160,171 @@
...
+          "type": "string",
+          "description": "The horizontal alignment of a node\u2019s label",
+          "enum": ["left", "center", "right"]
+        },
+        "text-valign": {
+          "type": "string",
+          "description": "The vertical alignment of a node\u2019s label",
+          "enum": ["top", "center", "bottom"]
+        },
+        "ghost": {
```
Comment (leecalcote) [inline schemas/constructs/v1beta1/component.json:265]: @aabidsofi19, can you offer justification as to why this should be an explicit property and not an interpretation of other properties, like `isAnnotation`?

Comment (aabidsofi19) [reply schemas/constructs/v1beta1/component.json:265]: @leecalcote  this is not ghost node property , but ghost is a style property that add shadow effect

File: schemas/constructs/v1alpha1/capability.json (hunk 11 above)
Comment (aabidsofi19) [reply schemas/constructs/v1alpha1/capability.json:18]: makes sense :+1:

**Review by aabidsofi19 — APPROVED** [review]

File: schemas/constructs/core.json [hunk 16]
```diff
@@ -36,6 +31,406 @@
...
+          "type": "number",
+          "description": "The opacity of the label text, including its outline.",
+          "minimum": 0,
+          "maximum": 1
+        },
+        "font-family": {
+          "type": "string",
+          "description": "A comma-separated list of font names to use on the label text."
+        },
+        "font-size": {
```
Comment (leecalcote) [inline schemas/constructs/core.json:76]: Intentionally, these are not camelCase, right? So long as this is intentional for UI client library compatibility, we're 👍

File: schemas/constructs/core.json [hunk 17]
```diff
@@ -36,6 +31,406 @@
...
+        "round-octagon",
+        "star",
+        "tag",
+        "round-tag",
+        "vee",
+        "polygon"
+      ]
+    },
+
+    "relationshipStyles": {
```
Comment (leecalcote) [inline schemas/constructs/core.json:141]: Relationships, not edges? @codeSafari10?

These styles apply to hierarchical, sibling, and other non-edge-based relationships?

File: schemas/constructs/core.json [hunk 18]
```diff
@@ -36,6 +31,406 @@
...
+      "type": "object",
+      "description": "Visualisation styles for a relationship",
```
Comment (leecalcote) [inline schemas/constructs/core.json:143]:
```suggestion
      "description": "Visualization styles for a relationship",
```

File: schemas/constructs/core.json [hunk 19]
```diff
@@ -36,6 +31,406 @@
...
+        "target-label": {
+          "type": "string",
+          "description": "The text to display for an edge\u2019s target label. Can give a path, e.g. data(id) will label with the elements id"
+        }
+      }
+    },
+
+    "componentStyles": {
+      "type": "object",
+      "description": "Visualisation styles for a component",
```
Comment (leecalcote) [inline schemas/constructs/core.json:255]:
```suggestion
      "description": "Visualization styles for a component",
```

File: schemas/constructs/core.json [hunk 20]
```diff
@@ -36,6 +31,406 @@
...
+          "description": "A CSS font weight to be applied to the label text."
+        },
+        "text-transform": {
+          "type": "string",
+          "description": "A transformation to apply to the label text",
+          "enum": ["none", "uppercase", "lowercase"]
+        },
+        "opacity": {
+          "type": "number",
+          "description": "The opacity of the element. See https://js.cytoscape.org/#style/visibility",
```
Comment (leecalcote) [inline schemas/constructs/core.json:95]:
```suggestion
          "description": "The opacity of the element, ranging from 0 to 1. Note that the opacity of a compound node parent affects the effective opacity of its children. See https://js.cytoscape.org/#style/visibility",
```

**Review by leecalcote — APPROVED** [review]
LGTM

File: schemas/constructs/v1alpha3/relationship.json [hunk 21]
```diff
@@ -28,11 +21,7 @@
     "kind": {
       "$ref": "../core.json#/definitions/inputString",
       "description": "Kind of the Relationship. Learn more about relationships - https://docs.meshery.io/concepts/logical/relationships.",
-      "enum": [
-        "Hierarchical",
-        "Edge",
-        "Sibling"
-      ]
+      "enum": ["Hierarchical", "Edge", "Sibling"]
```
Comment (leecalcote) [inline schemas/constructs/v1alpha3/relationship.json:24]: @MUzairS15, please confirm that lowercased versions of these enums are supported, too.

File: schemas/constructs/core.json (hunk 16 above)
Comment (aabidsofi19) [reply schemas/constructs/core.json:76]: yeah , for cytoscape compatibility .

File: schemas/constructs/core.json (hunk 17 above)
Comment (aabidsofi19) [reply schemas/constructs/core.json]: the generic set of styles apply to all , i will split the ones specif to edge

File: schemas/constructs/v1alpha3/relationship.json (hunk 21 above)
Comment (MUzairS15) [reply schemas/constructs/v1alpha3/relationship.json:24]: That would mean intercepting the validation, so let's keep them lowercase. Clients can format it when needed to display, ok?

Comment (leecalcote) [reply schemas/constructs/v1alpha3/relationship.json:24]: Sounds great. I will rely upon you to ensure that our current clients are compatible with lowercase values. Please suggest a change here and let’s merge and release.

Comment (leecalcote) [reply schemas/constructs/v1alpha3/relationship.json:24]: Relationship policies are currently compatible?

Comment (MUzairS15) [reply schemas/constructs/v1alpha3/relationship.json:24]: yes

Comment (MUzairS15) [inline schemas/constructs/v1alpha3/relationship.json:24]:
```suggestion
      "enum": ["hierarchical", "edge", "sibling"]
```

**Pull request merged**